### PR TITLE
Simplify handling of invalid signatures

### LINKED ISSUE #4111: Public key aggregation ambiguous infinite points handling
`BlstPublicKey.aggregate()` will return `infinitePublicKey` in two cases:

1. If one of the public keys to aggregate is invalid (not in the subgroup or infinity point)
2. If the sum of (valid) public keys happens to be zero

https://github.com/ConsenSys/teku/blob/a0270026d4e70db13b0042eb057ffbdb5709e4ca/bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstPublicKey.java#L56-L74

The two cases should be distinguished, so that the caller can handle them differently.

### PR DESCRIPTION
**Pull request by benjaminion**
The previous implementation used a flag to track whether a signature was valid or invalid. An example of an invalid signature was one that encodes a point not in the G2 group.

This led to oddities like having both valid and invalid infinity signatures. It also led to ambiguity about whether we check the G2 group membership at deserialisation or on usage of a signature. In batch verification we ended up doing both (but not explicitly checking the signature's validity in the calling code). This was all a bit messy and hard to maintain.

In this PR we explicitly delegate the group membership check to the Blst library, and we _do not_ perform the group membership check at deserialisation of the signature. This ought to slightly speed things up (due to no longer duplicating the check in batch verification) and should make things more maintainable.

The risk is that if Blst behaviour changes, or we swap in a different library in future, we could inadvertantly end up without the group membership check. I have added some tests to guard against this.

As a bonus, this closes #4111.

### CODE REVIEW

File: bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java [hunk 1]
```diff
@@ -17,34 +17,32 @@
...
   public static BlstSignature fromBytes(Bytes compressed) {
-    checkArgument(
-        compressed.size() == COMPRESSED_SIG_SIZE,
-        "Expected " + COMPRESSED_SIG_SIZE + " bytes of input but got %s",
-        compressed.size());
     try {
+      checkArgument(
+          compressed.size() == COMPRESSED_SIG_SIZE,
+          "Expected " + COMPRESSED_SIG_SIZE + " bytes of input but got %s",
+          compressed.size());
```
Comment (ajsutton) [inline bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java]: nit:
```suggestion
          "Expected %s bytes of input but got %s",
          COMPRESSED_SIG_SIZE,
          compressed.size());
```

File: bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java [hunk 2]
```diff
@@ -17,34 +17,32 @@
...
       P2_Affine ec2Point = new P2_Affine(compressed.toArrayUnsafe());
-      return new BlstSignature(ec2Point, ec2Point.in_group());
+      return new BlstSignature(ec2Point);
     } catch (Exception e) {
-      // Return an invalid infinite signature if deserialisation fails
-      return new BlstSignature(new P2_Affine(), false);
+      throw new DeserializeException("Signature deserialisation failed: " + e.getMessage());
```
Comment (ajsutton) [inline bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java]:
```suggestion
      throw new DeserializeException("Signature deserialization failed", e);
```

and add to `DeserializeException`:

```
  public DeserializeException(String message, Throwable cause) {
    super(message, cause);
  }
```

File: bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java [hunk 3]
```diff
@@ -57,21 +55,16 @@ static BlstSignature fromSignature(Signature signature) {
...
-      sum.add(finiteSignature.ec2Point);
+    try {
+      P2 sum = new P2();
+      for (BlstSignature finiteSignature : signatures) {
+        sum.aggregate(finiteSignature.ec2Point);
+      }
+      return new BlstSignature(sum.to_affine());
+    } catch (RuntimeException e) {
+      // Blst performs a G2 group membership test on each signature. We end up here if it fails.
+      throw new RuntimeException("Failed to aggregate signatures: " + e.getMessage());
```
Comment (ajsutton) [inline bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java]: I'm really not a fan of throwing a generic `RuntimeException` here, it's just too generic and you wind up catching `RuntimeException` all over the place then, potentially sweeping up exceptions you didn't intend to catch. I'd probably rename `DeserializeException` to `BlsException` so we can throw it here.  And worth keeping the full details of the original exception too with `new BlsException("Failed to aggregate signatures", e)`

**Review by ajsutton — COMMENTED** [review]
This generally looks good, I think it would be worth cleaning up how we're handling exceptions though so that we consistently throw a well-typed exception (probably rename `DeserializeException` to `BlsException`).  it would still extend from `RuntimeException` so be unchecked but makes handling it upstream so much easier.

File: bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java (hunk 1 above)
Comment (benjaminion) [reply bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java]: Done

File: bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java (hunk 2 above)
Comment (benjaminion) [reply bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java]: Done

File: bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java (hunk 3 above)
Comment (benjaminion) [reply bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java]: Good plan - done. I've reworked exceptions generally across the BLS classes. It's not perfect, but simpler than it was.

File: bls/src/main/java/tech/pegasys/teku/bls/BLS.java [hunk 4]
```diff
@@ -141,21 +146,25 @@ public static BLSSignature aggregate(List<BLSSignature> signatures)
...
-                (pk, msg) -> new PublicKeyMessagePair(pk.getPublicKey(), msg))
-            .collect(Collectors.toList());
     try {
-      return signature.getSignature().verify(publicKeyMessagePairs);
-    } catch (DeserializeException e) {
-      return false;
+      checkArgument(
+          publicKeys.size() == messages.size(),
+          "Number of public keys and number of messages differs.");
+      if (publicKeys.isEmpty()) return false;
```
Comment (ajsutton) [inline bls/src/main/java/tech/pegasys/teku/bls/BLS.java]: nit: We should be using braces on this if statement.

File: bls/src/main/java/tech/pegasys/teku/bls/BLS.java [hunk 5]
```diff
@@ -178,13 +187,18 @@ public static boolean fastAggregateVerify(
...
     }
-    if (publicKeys.isEmpty()) return false;
-    List<PublicKey> publicKeyObjects =
-        publicKeys.stream().map(BLSPublicKey::getPublicKey).collect(Collectors.toList());
+
     try {
-      return signature.getSignature().verify(publicKeyObjects, message);
-    } catch (DeserializeException e) {
-      return false;
+      if (publicKeys.isEmpty()) return false;
```
Comment (ajsutton) [inline bls/src/main/java/tech/pegasys/teku/bls/BLS.java]: nit: Braces.

**Review by ajsutton — APPROVED** [review]
LGTM.

File: bls/src/main/java/tech/pegasys/teku/bls/BLS.java (hunk 5 above)
Comment (benjaminion) [reply bls/src/main/java/tech/pegasys/teku/bls/BLS.java]: Haha, I didn't perpetrate this :-)  But have fixed it and a couple of others.

**Pull request merged**